napari/layers/utils/text.py

```diff
@@ -1,242 +1,96 @@
+    properties: dict
+        Stores properties data that will be used to generate strings when text
+        is a format a string.
 
     Attributes
     ----------
-    text : array
-        the strings to be displayed
-    rotation : float
-        Angle of the text elements around the anchor point. Default value is 0.
-    anchor : str
+    values : np.ndarray
+        The text values to be displayed.
+    visible : bool
+        True if the text should be displayed, false otherwise.
+    size : float
+        Font size of the text, which must be positive. Default value is 12.
+    color : array
+        Font color for the text as an [R, G, B, A] array.
+    blending : Blending
+        The blending mode that determines how RGB and alpha values of the layer
+        visual get mixed. Allowed values are {'opaque', 'translucent', and 'additive'}.
+        Note that 'opaque` blending is not recommended, as it colors the bounding box
+        surrounding the text.
+    anchor : Anchor
         The location of the text origin relative to the bounding box.
         Should be 'center', 'upper_left', 'upper_right', 'lower_left', or 'lower_right'.
     translation : np.ndarray
         Offset from the anchor point.
-    color : array
-        Font color for the text
-    size : float
-        Font size of the text. Default value is 12.
-    The blending mode that determines how RGB and
-        alpha values of the layer visual get mixed. Allowed values are
-        {'opaque', 'translucent', and 'additive'}. Note that 'opaque` blending
-        is not recommended, as colors the bounding box surrounding the text.
-    visible : bool
-        Set to true of the text should be displayed.
+    rotation : float
+        Angle of the text elements around the anchor point. Default value is 0.
     """
 
-    def __init__(
-        self,
-        text,
-        n_text,
-        properties={},
-        rotation=0,
-        translation=0,
-        anchor='center',
-        color='cyan',
-        size=12,
-        blending='translucent',
-        visible=True,
-    ):
-
-        self.events = EmitterGroup(
-            source=self,
-            auto_connect=True,
-            text=Event,
-            rotation=Event,
-            translation=Event,
-            anchor=Event,
-            color=Event,
-            size=Event,
-            blending=Event,
-            visible=Event,
-        )
-
-        self.events.block_all()
-        self._rotation = rotation
-        self._anchor = Anchor(anchor)
-        self._translation = translation
-        self._color = transform_color(color)[0]
-        self._size = size
-        self._blending = self._check_blending_mode(blending)
-        self._visible = visible
-
-        self._set_text(text, n_text, properties)
-        self.events.unblock_all()
-
-    @property
-    def values(self):
-        """np.ndarray: the text values to be displayed"""
-        return self._values
+    values: np.ndarray = None
```

Comment:
You might want to use our custom array type here https://github.com/napari/napari/blob/master/napari/utils/events/custom_types.py#L4

It lets you place additional constraints on the shapes, and I think dtype of the array. You can see how we use it in the colormap here https://github.com/napari/napari/blob/d310847f0c79f062daadecb8a159e8972c074470/napari/utils/colormaps/colormap.py#L53

Reply:
Done and agreed that it's a nice small improvement.

I was hoping this would mean that I could also remove some of the more boring pydantic validators here, but it didn't help. If I try the following and remove the validator that assigns a default if `values` is None:

```python
values: Array[str, (-1,)] = np.empty(0)
```

then I get the following deprecation error + stack trace:

```
    class TextManager(EventedModel):
../../../utils/events/evented_model.py:79: in __new__
    cls = super().__new__(mcs, name, bases, namespace, **kwargs)
pydantic/main.py:299: in pydantic.main.ModelMetaclass.__new__
    ???
pydantic/fields.py:411: in pydantic.fields.ModelField.infer
    ???
pydantic/fields.py:342: in pydantic.fields.ModelField.__init__
    ???
pydantic/fields.py:445: in pydantic.fields.ModelField.prepare
    ???
pydantic/fields.py:473: in pydantic.fields.ModelField._set_default_and_type
    ???
pydantic/fields.py:345: in pydantic.fields.ModelField.get_default
    ???
pydantic/utils.py:627: in pydantic.utils.smart_deepcopy
    ???
E   DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
```

It looks like pydantic is relying on being able to evaluate a numpy array as a truthy value when making a default instance, but that was deprecated by numpy a few years ago and now errors. Using `[]` as a default value works, but then the value being stored doesn't actually conform to the type, which seems confusing.

So for optional attributes that store numpy arrays, is it OK and recommended to implemented a validator to handle default values?

Reply:
One minor issue with using `Array` instead of `np.ndarray` is that my IDE (fairly) complains about a type mismatch (i.e. it expects the subtype `Array`, but we give it the less specific `np.ndarray`). I've seen this in a few other places in the code and it's a only minor distraction, but just wondering if there are any ideas on how to fix that (either internally or in pydantic).

Reply:
> So for optional attributes that store numpy arrays, is it OK and recommended to implemented a validator to handle default values?

@tlambert03 do you have additional thoughts on this? I think code that is there now looks great to me - matches what we've don in other places

> One minor issue with using Array instead of np.ndarray is that my IDE (fairly) complains about a type mismatch 

Good to know. @tlambert03 might have thoughts here too. Could be something we address elsewhere too

Reply:
been trying to catch up on some other things here today... but definitely want to give this a close look!  will do by the end of the day